## 예외
자바 개발자가 가장 신경 쓰기 귀찮아하는 것 중의 하나가 **예외처리**다. 3장에서 JdbcTemplate을 적용하면서 deleteAll() 메소드나 add() 메소드에 ```throws SQLException```이 없어졌다.
왜 달라졌는지 살펴보자.

### 예외 블랙홀
자바로 프로그래밍을 하다보면 예외가 일어날 수 있는 코드를 그냥 작성하면 IDE에서 친절하게 빨간 줄을 그어주며 '처리되지 않은 예외가 있다'고 에러 표시를 해준다. 이를 보고 try/catch문을 해당 코드를 감싸면 별문제 없이 잘 동작한다. 
예외가 발생할 때 그것을 catch 블록을 통해 잡아내는 것까지는 좋은데 안에 비워두고 그냥 넘어가 버리는 것은 매우 위험한 일이다. 왜냐하면 프로그램 실행 중에 어된 디선가 오류가 있어 예외가 발생했는데 그것을 무시하고 계속 진행해 버리기 때문이다. 이는 나중에 더 큰 문제를 일으킬 수 있고 원인이 무엇인지 찾아내기가 매우 어려워질 수 있는 위험이 있다.  
경우에 따라 **의도된 목적**으로 비워두는 경우를 제외하고는 빈 catch 블록으로 예외 처리하는 것을 해선 안된다.  
~~~java 
catch(SQLException e){
  System.out.println(e);
}

catch(SQLException e){
  e.printStackTrace();
}
~~~
또는 위에처럼 단순히 예외를 출력해주고 넘어가버리는 것도 좋지 않은 방법이다. 예외를 처리할 때 반드시 지켜야 할 핵심 원칙은 한가지다. **모든 예외는 적절하게 복구되든지 아니면 작업을 중단시키고 운영자 또는 개발자에게 분명하게 통보돼야 한다.**  
만약 예외를 잡아서 뭔가 조치를 취할 방법이 없다면 ```throws SQLException```을 선언해서 예외를 메소드 밖으로 던져 자신을 호출한 코드에 예외처리 책임을 전가하는 것도 방법이다.  
~~~java
public void method1() throws Exception {
  method2();
}
public void method2() throws Exception {
  method3();
}

  ...
  
public void method10() throws Exception {
  method11();
}
~~~
그렇다고 예외들을 자기를 호출하는 코드에 계속 throw 해버리면 위와 같은 상황이 발생한다. 이렇게 무책임하게 throw 해버리는 것은 적절한 처리를 통해 복구될 수 있는 예외상황도 제대로 다룰 수 있는 기회를 박탈당한다.  
위에 나오는 두 가지 나쁜 습관은 어떤 경우에도 용납하지 않아야 한다.

### 예외의 종류와 특징
자바에서 throw를 통해 발생시킬 수 있는 예외는 크게 3가지가 있다.
+ Error
    java.lang.Error 클래스의 서브 클래스들이다. 에러는 시스템에 비정상적인 상황이 발생했을 경우에 사용된다. 이 경우, 애플리케이션 코드에서 에러를 잡아도 아무런 대응을 할 수 없기 때문에 시스템 레벨에서 특별한 작업을 하는게 아니라면 애플리케이션에서는 이런 에러에 대한 처리는 신경쓰지 않아도 된다.  
+ Exception과 체크 예외
    java.lang.Exception 클래스와 그 서브클래스들이다. 개발자들이 만든 애플리케이션 코드의 작업 중에 예외 상황이 발생 했을 경우에 사용된다.  
    Exception 클래스는 다시 체크 예외와 언체크 예외로 구분된다. 체크 예외는 Exception 클래스의 서브클래스 이면서 RuntimeException 클래스를 상속하지 않은 것들이고, 언체크 예외는 RuntimeException을 상속한 클래스들을 말한다.  
+ RuntimeException과 언체크/런타임 예외
    java.lang.RuntimeException 클래스를 상속한 예외들은 명시적인 예외처리를 하지 않기 때문에 언체크 예외라고 불린다. 에러와 마찬가지로 catch문으로 예외를 잡거나 throws를 선언하지 않아도 된다.  
    런타임 예외는 주로 프로그램의 오류가 있을 때 발생하도록 의도된 것들이다. 피할 수 있지만 개발자가 부주의해서 발생할 수 있는 경우에 발생하도록 만든 것이 런타임 예외이기 대문에 굳이 catch나 throws를 사용하지 않아도 된다.  
  
  

### 예외처리 방법

#### 예외 복구
예외 복구는 예외 상황을 파악하고 문제를 해결해서 정상 상태로 돌려놓는 것이다. 서비스 흐름에서 예외가 발생했을 때 다른 작업 흐름으로 자연스럽게 유도해주는 것이다. 단순히 사용자에게 에러 메시지만 던지는 것은 예외 복구라고 할 수 없다. 예외가 처리됐으면 비록 기능적으로는 사용자에게 예외상황으로 비쳐도 애플리케이션에서는 정상적으로 설계된 흐름을 따라 진행돼야 한다.  
만약 네트워크 접속이 원활하지 않아서 예외가 발생했다면 일정 시간 대기했다가 다시 접속을 시도해보는 방법을 사용해서 예외상황으로부터 복구를 시도할 수 있다. 물론 정해진 횟수만큼 재시도해서 실패했다면 예외 복구는 포기해야 한다.  
  
   
#### 예외처리 회피
예외처리 회피는 예외처리를 자신이 담당하지 않고 자신을 호출한 쪽으로 던져버리는 것이다. ```throws```를 사용하거나 ```catch```블록으로 일단 예외를 잡은 후에 로그를 남기고 다시 예외를 던지는(throw) 것이다. 다시 말해 예외를 자신이 처리하지 않고 회피하는 방법이다.  
예외를 회피하는 것은 예외를 복구하는 것처럼 의도가 분명해야 한다. 템플릿/콜백처럼 긴밀한 관계에 있는 다른 오브젝트에게 예외처리 책임을 분명히 지게 하거나, 자신을 사용하는 쪽에서 예외를 다루는게 최선의 방법이라는 분명한 확신이 있어야 한다.  
  
  
#### 예외 전환
예외 전환은 **예외처리 회피**와 비슷하게 예외를 메소드 밖으로 던지는 것이다. 하지만 예외 회피와는 달리, 발생한 예외를 그대로 넘기는 것이 아니라 적절한 예외로 전환해서 던진다는 특징이 있다. 예외 전환은 보통 두 가지 목적으로 사용된다.  
  
  
  
첫번째 전환 방법은 내부에서 발생한 예외를 그대로 던지는 것이 그 예외상황에 대한 적절한 의미를 부여해주지 못하는 경우에, 의미를 분명하게 해줄 수 있는 예외로 바꿔주기 위해서다.
예를 들어, 새로운 사용자를 등록하려고 시도했을 때 아이디가 같은 사용자가 있으면 SQLException을 발생한다. 이 경우에 단순히 SQLException을 던져버리면 서비스 계층에서는 해당 에러가 아이디 중복으로 인한 에러인지 DB서버에 장애가 있는지 쉽게 구분하기 어렵다.  
의미가 분명한 예외를 던지면 서비스 계층 오브젝트가 적절한 복구 작업을 시도할 수 있다.
~~~java
public void add(User user) throws DuplicateUserException, SQLException {
  try{
    ... //user를 DB에 추가하는 코드
  }catch(SQLException e){
    if(e.getErrorCode == MysqlErrorNumbers.ER_DUP_ENTRY){
      throw DuplicateUserException();
    }else{
      throw e;
    }
  }

}
~~~
 
 
두번째 전환 방법은 예외를 처리하기 쉽고 단순하게 만들기 위해 포장하는 것이다. 첫번째 방법과 달리 예외의 의미를 명확하게 하려 다른 예외로 전환하는 것이 아니다. 주로 예외처리를 강제하는 체크 예외를 언체크 예외인 런타임 예외로 바꾸는 경우에 사용한다. 
일부 코드에서 발생하는 예외는 복구가 불가능한 경우가 있다. 이 경우 런타임 예외로 포장해서 던지고 자세한 로그를 남기고, 관리자에게 메일 등으로 통보해주는 것이 낫다.  
  
### JdbcTemplate의 예외처리
4장 첫부분에 deleteAll() 메소드나 add() 메소드에 ```throws SQLException```이 없어졌는지 의문을 던졌었다. SQLException은 99% 코드 레벨에서 복구할 방법이 없다. 프로그램의 오류 또는 개발자의 부주의 때문에 발생하는 경우다.  
그렇기 때문에 스프링의 JdbcTemplate은 템플릿과 콜백 안에서 발생하는 모든 SQLException을 런타임 예외인 DataAccessException으로 포장해서 던진다. 따라서, JdbcTemplate을 사용하는 Dao 메소드에서는 꼭 필요한 경우에만 런타임 예외인 DataAccessException을 잡아서 처리하면 된다.  
  
  
